features/79-structured-logging/structured-logging.md

```diff
@@ -0,0 +1,69 @@
+# Structured Logging
+
+## Purpose
+
+To make log output structured so that the log output can be filtered or
+otherwise processed.
+
+### Log output
+
+### Common structured log data
+
+* process - process name
+* logger - name of the logger that generates the log entry
+* datetime - datetime when the log entry is emitted
+* level - indicate the level of severity for the log entry
+* message - log message
+
+### More log data
+
+The following items are not necessarily available in all situations, but
+they should be part of the structured log data if available.
+
+* request_id - request ID, generated by router
+
+### Logger
+
+Logger is the primary field for filtering log entries.
+
+Proposing the following list of loggers:
+
+* server - skygear-server log not otherwise covered by other loggers
+* server.sql - skygear-server log for executed sql statement and result
+* server.http - skygear-server log for http request and response body
+* server.transport skygear-server log for plugin transport
+* plugin - plugin log not otherwise covered by other loggers
+* plugin.transport - plugin log for plugin transport
+* cloudfunc - cloud function log generated by developer code
+
+### Log format
+
+Will adopt JSON log format.
+
+## Configuration
+
+TBD: Skygear-server can be configured to turn off logs based on logger and
+level.
+
+## Cloud App Image
+
+Cloud app image should print log entries to stdout for Kubernetes to collect.
+The cloud app image should print in JSON format. To do that,
+
+* Processes stdout/stderr is redirected a log processor running in the
+  container (e.g. Run `skygear-server | log-processor`).
+* Log processor to check if each log entry is in JSON format.
+* If log entry is not in JSON format, wrap the log entry in JSON format with
+  common stuctured log data.
+* Log processor to print log to container stdout.
+
+## Portal
+
+* Print JSON-formatted log entries in a nice format (table or otherwise)
+* Support filtering using the `logger`
+* Colorize log output based on `logger` (or other fields)
+* Filter log output by selecting interested `logger` (likely multi-select)
```

Comment:
Seem we will want a wireframe to process with the portal. So the scope will be clear.